src/core/server/config/config_service.ts

```diff
@@ -33,19 +41,30 @@ export type IConfigService = PublicMethodsOf<ConfigService>;
 export class ConfigService {
   private readonly log: Logger;
 
+  private readonly config$: Observable<Config>;
+
   /**
    * Whenever a config if read at a path, we mark that path as 'handled'. We can
    * then list all unhandled config paths when the startup process is completed.
    */
   private readonly handledPaths: ConfigPath[] = [];
   private readonly schemas = new Map<string, Type<unknown>>();
+  private readonly deprecations = new BehaviorSubject<ConfigDeprecationWithContext[]>([]);
 
   constructor(
-    private readonly config$: Observable<Config>,
+    private readonly rawConfigProvider: RawConfigurationProvider,
     private readonly env: Env,
     logger: LoggerFactory
   ) {
     this.log = logger.get('config');
+
+    this.config$ = combineLatest([this.rawConfigProvider.getConfig$(), this.deprecations]).pipe(
+      map(([rawConfig, deprecations]) => {
+        const migrated = applyDeprecations(rawConfig, deprecations);
+        return new LegacyObjectToConfigAdapter(migrated);
```

Comment:
The `LegacyObjectToConfigAdapter` transformation probably needs to be applied before the deprecations, as we want to use the legacy-adapted values for deprecations. Will need to change the implementation of the adapter though, as it now needs to returns a `Record` instead of a config.

BTW `LegacyObjectToConfigAdapter` is quite a dangerous implementation, as it currently only mutates the values during access.

https://github.com/elastic/kibana/blob/033193b52a0bebb350067f8ca09ecded85d757f7/src/core/server/legacy/config/legacy_object_to_config_adapter.ts#L88-L100